design/tlog-spilling.md.html

```diff
@@ -0,0 +1,590 @@
+transaction log.  The old generation of transaction logs will only shut down
+once all the data that they have has been fully popped.  This means that there
+can be multiple instances of a transaction log 
+
+*********************************************************
+*                   SharedTLog                          *
+*                                                       *
+* +--------+--------+--------+--------+--------+        *
+* | TLog 1 | TLog 2 | TLog 3 | TLog 4 | TLog 5 |        *
+* +--------+--------+--------+--------+--------+        *
+*   ^ popping         ^spilling         ^committing     * 
+*********************************************************
+
+
+
+
+### Transaction State Store
+
+For FDB to perform a recovery, there is information that it needs to know about
+the database, such as the configuration, worker exclusions, backup status, etc.
+These values are stored into the database in the `\xff` system keyspace.
+However, during a recovery, FDB can't read this data from the storage servers,
+because recovery hasn't completed, so it doesn't know who the storage servers
+are yet.  Thus, a copy of this data is held in-memory on every proxy in the
+*transaction state store*, and durably persisted as a part of commits on the
+transaction logs.  Being durably stored on the transaction logs means the list
+of transaction logs can be fetched from the coordinators, and then used to load
+the rest of the information about the database.
+
+The in-memory storage engine writes an equal amount of mutations and snapshot
+data to a queue, an when a full snapshot of the data has been written, deletes
+the preceeding snapshot and begins writing a new one.  When backing an
+in-memory storage engine with the transaction logs, the
+`LogSystemDiskQueueAdapter` implements writing to a queue as committing
+mutations to the transaction logs with a special tag of `txsTag`, and deleting
+the preceeding snapshot as popping the transaction logs for the tag of `txsTag`
+until the version where the last full snapshot began.
+
+This means that unlike every other commit that is tagged and stored on the
+transaction logs, `txsTag` signifies data that is:
+
+1. Committed to infrequently
+2. Only peeked on recovery
+3. Popped infrequently, and a large portion of the data is popped at once
+4. A small total volume of data
+
+The most problematic of these is the infrequent popping.  Unpopped data will be
+spilled after some time, and if `txsTag` data is spilled and not popped, it
+will prevent the DiskQueue from being popped as well.  This will cause the
+DiskQueue to grow continuously.  The infrequent commits and small data volume
+means that there benefits of spill-by-reference over spill-by-value don't apply
+for this tag.
+
+Thus, even when configured to spill-by-reference, `txsTag` is spilled by value.
+
+### Disk Queue Recovery
+
+If a transaction log dies and restarts, all commits that were in memory at the
```

Comment:
Makes sense but I would also move this paragraph up towards the top to make it clear right at the beginning what the DiskQueue is currently used for so that when you start talking about the new stuff its easier to contrast the new behavior (which requires a random seek API) to the old behavior (which was only ever read start to finish when the TLog node restarted)